pilot/pkg/proxy/envoy/v2/debounced_channel.go

```diff
@@ -0,0 +1,143 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v2
+
+import (
+	"sync"
+	"time"
+)
+
+// DebouncedEvent is used as entry type, which is enqueued in DebouncedChannel
+type DebouncedEvent = bool
+
+// DebouncedChannel is a channel which abstracts the debounce algorithm
+type DebouncedChannel struct {
+	// C channel, which can be used in selects. It delivers debounced events.
+	// The function emitted must be called. Otherwise processing is stopped.
+	C <-chan func(func(DebouncedEvent))
+	c chan func(func(DebouncedEvent))
+
+	pushCounter     uint32
+	debounceCounter uint32
+	debounceEndAt   time.Time
+	lastEnqueueTime time.Time
+	lastEmitTime    time.Time
+	debounceAfter   time.Duration
+	debounceMax     time.Duration
+	mutex           sync.Mutex
+
+	merge              func(accumulator DebouncedEvent, value DebouncedEvent) DebouncedEvent
+	accumulator        DebouncedEvent
+	initialAccumulator DebouncedEvent
+
+	running bool
+}
+
+// NewDebouncedChannel create a new instance of DebouncedChannel
+// debounceAfter: A new outgoing event is emitted if no additional event is enqueued during this period (resets on every enqueued event).
+// debounceMax: A new outgoing event is emitted after this period (even if additional events were enqueued in that period).
+// initialAccumulator: Initial value of the accumulator
+// merge: function, which is called to merge two events if they are debounced.
+func NewDebouncedChannel(debounceAfter time.Duration, debounceMax time.Duration,
+	initialAccumulator DebouncedEvent, merge func(accumulator DebouncedEvent, value DebouncedEvent) DebouncedEvent) *DebouncedChannel {
+
+	c := make(chan func(func(DebouncedEvent)), 1)
+
+	return &DebouncedChannel{
+		C:                  c,
+		c:                  c,
+		initialAccumulator: initialAccumulator,
+		accumulator:        initialAccumulator,
+		merge:              merge,
+		debounceAfter:      debounceAfter,
+		debounceMax:        debounceMax,
+		lastEmitTime:       time.Unix(0, 0),
+	}
+
+}
+
+// Enqueue pushes a new event to the channel
+func (dc *DebouncedChannel) Enqueue(element DebouncedEvent) {
+	dc.mutex.Lock()
+	defer dc.mutex.Unlock()
+	dc.accumulator = dc.merge(dc.accumulator, element)
+	if dc.debounceCounter == 0 {
+		go func() {
+			dc.mutex.Lock()
+			defer dc.mutex.Unlock()
+
+			maxWaitUntil := time.Now().Add(dc.debounceMax)
+			for {
+				debounceCounter := dc.debounceCounter
+
+				dc.mutex.Unlock()
+				time.Sleep(min(dc.debounceAfter, time.Until(maxWaitUntil)))
```

Comment:
I think there is a race here? Consider:

* Enqueue 1, it gets to this sleep.
* Enqueue 2, the mutex is open so it goes through everything, calling runIfNeeded, triggering push
* Enqueue 1 wakes up, calls runIfNeeded again, also triggering a push

Reply:
No the second Enqueue will not enter the if condition above (`debounceCounter == 0`) which is in a locked section. For the second Enqueue, the `debounceCounter` is 1.

Reply:
yeah but outside the if there is another runIfNeeded right? I'm going to pull this down locally and try it out, hard to tell just from the GitHub ui

Reply:
Ok I see.
`runIfNeeded` is always called with a locked mutex. The push is only trigger once, because the push only happens if the `debounceCounter` greater than zero. The `debounceCounter` is reset to zero in this case.
The second call to `runIfNeeded` will not enter the `if` condition.